From e2bcdb9c2c27e9e6e985138b663c871bc36c4b19 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Mon, 4 May 2026 16:06:25 -0700
Subject: [PATCH] Prevent SQL injection in SelectionBuilder

Use SQLiteTokenizer.OPTION_CHECK_BRACKETS to ensure that selection
clauses added to SelectionBuilder do not have unbalanced brackets.
This prevents callers from breaking out of the parentheses used
when concatenating clauses.

Bug: 393377017
Test: atest SelectionBuilderTest
Flag: EXEMPT BUGFIX

Autogenerated by MendIt (go/androidmendit)

Change-Id: If178bd2ab556147d7e86771e6157c531499d8ff5
---
 .../contacts/util/SelectionBuilder.java         |  2 ++
 .../contacts/util/SelectionBuilderTest.java     | 17 +++++++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/src/com/android/providers/contacts/util/SelectionBuilder.java b/src/com/android/providers/contacts/util/SelectionBuilder.java
index 14499e8..e21e2e2 100644
--- a/src/com/android/providers/contacts/util/SelectionBuilder.java
+++ b/src/com/android/providers/contacts/util/SelectionBuilder.java
@@ -16,6 +16,7 @@
 
 package com.android.providers.contacts.util;
 
+import android.database.sqlite.SQLiteTokenizer;
 import android.text.TextUtils;
 
 import java.util.ArrayList;
@@ -44,6 +45,7 @@ public class SelectionBuilder {
      */
     public SelectionBuilder addClause(String clause) {
         if (!TextUtils.isEmpty(clause)) {
+            SQLiteTokenizer.tokenize(clause, SQLiteTokenizer.OPTION_CHECK_BRACKETS, null);
             mWhereClauses.add(clause);
         }
         return this;
diff --git a/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java b/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java
index 06f7e29..5d96c90 100644
--- a/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java
+++ b/tests/src/com/android/providers/contacts/util/SelectionBuilderTest.java
@@ -54,4 +54,21 @@ public class SelectionBuilderTest extends FixedAndroidTestCase {
         assertEquals("(A)", new SelectionBuilder("").addClause("A").build());
         assertEquals("(A) AND (B)", new SelectionBuilder("A").addClause("B").build());
     }
+
+    public void testInjection() {
+        try {
+            new SelectionBuilder("1=1)) OR ((1=1");
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            // expected
+        }
+
+        try {
+            SelectionBuilder sb = new SelectionBuilder("A");
+            sb.addClause("1=1)) OR ((1=1");
+            fail("Should have thrown");
+        } catch (IllegalArgumentException expected) {
+            // expected
+        }
+    }
 }
-- 
2.55.0

